chrome/content/zotero/xpcom/storage/storageCache.js

```diff
@@ -0,0 +1,319 @@
+	 *
+	 * @returns {Promise<void>}
+	 */
+	identifyItemsToFree: async function (forGroups) {
+		this._freeingCache = true;
+		let promises = [];
+		try {
+			if (forGroups) {
+				Zotero.Libraries.getAll().forEach((library) => {
+					if (library.libraryID !== Zotero.Libraries.userLibraryID) {
+						promises.push(
+							this.identifyItemsToFreeForLibrary(library.libraryID)
+						);
+					}
+				});
+			}
+			else {
+				promises.push(
+					this.identifyItemsToFreeForLibrary(Zotero.Libraries.userLibraryID)
+				);
+			}
+	
+			await Zotero.Promise.all(promises);
+		}
+		catch (e) {
+			throw e;
+		}
+		finally {
+			this._freeingCache = false;
+		}
+	},
+
+	/**
+	 * Identify items with files currently downloaded that we should delete to limit cache
+	 *
+	 * Called when the limit cache preference is turned on or when a new item is downloaded or
+	 * added.
+	 *
+	 * @param libraryID {integer} - Library ID to search for items to free
+	 *
+	 * @returns {Promise<void>}
+	 */
+	identifyItemsToFreeForLibrary: async function (libraryID) {
+		// Check library has storage
+		if (!Zotero.Sync.Storage.Local.getEnabledForLibrary(libraryID)) {
+			Zotero.debug('Zotero.Sync.Storage.Cache ('
+				+ libraryID + ') exiting because it is not storage enabled ('
+				+ Zotero.Libraries.get(libraryID).libraryType + ').');
+			return;
+		}
+
+		// Check this library is on demand
+		if (!Zotero.Sync.Storage.Local.downloadAsNeeded(libraryID)) {
+			Zotero.debug('Zotero.Sync.Storage.Cache (' + libraryID
+				+ ') exiting because it is not download as-needed.');
+			return;
+		}
+		
+		let cacheTime = Zotero.Prefs.get(this._getCacheLimitPrefFromLibrary(libraryID));
+		if (cacheTime === 0) {
+			Zotero.debug('Zotero.Sync.Storage.Cache (' + libraryID
+				+ ') exiting because cache time is unlimited.');
+			return;
+		}
+
+		Zotero.debug('Zotero.Sync.Storage.Cache (' + libraryID
+			+ ') checking if cached files exceeds preference of ' + cacheTime + ' days.');
+
+		// Turn cache limit into seconds for timestamp
+		cacheTime = cacheTime * 24 * 60 * 60;
+
+		// Keep track of offset into table
+		let offset = 0;
+
+		// Delete files older than the cache limit
+		let lastAccessedIsNull = false, i, recordsFreed = 0;
+		while (true) {
+			let sql = "SELECT itemID FROM itemAttachments JOIN items USING (itemID) "
+				+ " WHERE libraryID=? AND linkMode IN (?,?) AND syncState IN (?)";
+			
+			// If we have no last accessed data in DB then compare with dateModified
+			if (lastAccessedIsNull) {
+				sql += " AND lastAccessed IS NULL AND dateModified < ?";
+			}
+			else {
+				sql += " AND lastAccessed IS NOT NULL AND lastAccessed > ?";
+			}
+			
+			sql += " ORDER BY lastAccessed DESC, dateModified DESC LIMIT ?,1000";
+			let params = [
+				libraryID,
+				Zotero.Attachments.LINK_MODE_IMPORTED_FILE,
+				Zotero.Attachments.LINK_MODE_IMPORTED_URL,
+				Zotero.Sync.Storage.Local.SYNC_STATE_IN_SYNC,
+				cacheTime,
+				offset
```

Comment:
Is using an offset correct here? If you're changing syncState when deleting the files, won't they not be matched in subsequent queries, such that an offset would skip thousands of unprocessed files?